# Predictive Maintenance Analysis For Industrial Equipment

**Project Overview -**
This project analyzes industrial equipment data to identify failure patterns and predict potential machine breakdowns. The goal is to help manufacturing companies — particularly in the **German and European industrial sector** — reduce unplanned downtime, optimize maintenance schedules, and cut operational costs.

## Problem Statement
In modern manufacturing, unplanned equipment failures are one of the biggest threats to productivity. Every hour of downtime costs companies thousands of euros in lost production, emergency repairs, and supply chain disruption.

## Project Goals
| # | Goal | Method |
|---|------|---------|
| 1 | Identify failure-causing factors | Correlation & statistical analysis |
| 2 | Classify common failure types | Categorical analysis & visualization |
| 3 | Detect failures in advance | Time-based pattern analysis |
| 4 | Optimize maintenance schedule | Cost-benefit analysis & recommendations |

## Notebook - 01_data_exploration

**Goal:** Load the dataset, understand its structure, and perform initial exploration.

**Steps:**
1. Import Libraries
2. Load Data
3. Dataset Shape & Structure
4. Data Types & Missing Values
5. Basic Statistical Summary
6. Categorical Column Exploration
7. Target Variable Analysis (Machine Failure)
8. Failure Type Breakdown
9. Initial Observations & Next Steps

## 1. IMPORT LIBRARIES

In [1]:
# Import Libraries
import os
os.chdir('..')
print(os.getcwd())

import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')
print("All Libraries Import Successfully")

/Users/jayeshranghera/Documents/Projects/predictive-maintenance-analysis
All Libraries Import Successfully


## 2. LOAD DATASET

In [8]:
df = pd.read_csv('data/raw/ai4i2020.csv')


## 3. Shape and Structure

In [10]:
print(f'Dataset Shape: {df.shape}')
df.head()     # first 5 raws

Dataset Shape: (10000, 14)


,UDI,Product ID,Type,Air temperature [K],Process temperature [K],Rotational speed [rpm],Torque [Nm],Tool wear [min],Machine failure,TWF,HDF,PWF,OSF,RNF
0,1,M14860,M,298.1,308.6,1551,42.8,0,0,0,0,0,0,0
1,2,L47181,L,298.2,308.7,1408,46.3,3,0,0,0,0,0,0
2,3,L47182,L,298.1,308.5,1498,49.4,5,0,0,0,0,0,0
3,4,L47183,L,298.2,308.6,1433,39.5,7,0,0,0,0,0,0
4,5,L47184,L,298.2,308.7,1408,40.0,9,0,0,0,0,0,0


In [12]:
# Column Names
print(f' Column List: {df.columns.tolist()}')

 Column List: ['UDI', 'Product ID', 'Type', 'Air temperature [K]', 'Process temperature [K]', 'Rotational speed [rpm]', 'Torque [Nm]', 'Tool wear [min]', 'Machine failure', 'TWF', 'HDF', 'PWF', 'OSF', 'RNF']


In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   UDI                      10000 non-null  int64  
 1   Product ID               10000 non-null  object 
 2   Type                     10000 non-null  object 
 3   Air temperature [K]      10000 non-null  float64
 4   Process temperature [K]  10000 non-null  float64
 5   Rotational speed [rpm]   10000 non-null  int64  
 6   Torque [Nm]              10000 non-null  float64
 7   Tool wear [min]          10000 non-null  int64  
 8   Machine failure          10000 non-null  int64  
 9   TWF                      10000 non-null  int64  
 10  HDF                      10000 non-null  int64  
 11  PWF                      10000 non-null  int64  
 12  OSF                      10000 non-null  int64  
 13  RNF                      10000 non-null  int64  
dtypes: float64(3), int64(9)

## 4.Data Types & Missing Values

In [24]:
missing_df = pd.DataFrame({
    'columns': df.columns,
    'Data Type': df.dtypes.values,
    'Missing Values': df.isnull().sum().values,
    'Missing %': ((df.isnull().sum() / len(df))*100).round(2).values
})

print("Data Types and Missing Values")
missing_df

Data Types and Missing Values


,columns,Data Type,Missing Values,Missing %
0,UDI,int64,0,0.0
1,Product ID,object,0,0.0
2,Type,object,0,0.0
3,Air temperature [K],float64,0,0.0
4,Process temperature [K],float64,0,0.0
5,Rotational speed [rpm],int64,0,0.0
6,Torque [Nm],float64,0,0.0
7,Tool wear [min],int64,0,0.0
8,Machine failure,int64,0,0.0
9,TWF,int64,0,0.0


In [31]:
# Duplicate rows check
duplicates = df.duplicated().sum()
print(f'Duplicate Rows: {duplicates}')

if duplicates == 0:
    print("Yehh - No Duplicate rows found.")
else:
    print("Duplicate Rows Found - Need Attention")

Duplicate Rows: 0
Yehh - No Duplicate rows found.


## 5.Basic Statistical Summary

In [35]:
print(' Numerical Feature Summery:')
df.describe().round(2)

 Numerical Feature Summery:


,UDI,Air temperature [K],Process temperature [K],Rotational speed [rpm],Torque [Nm],Tool wear [min],Machine failure,TWF,HDF,PWF,OSF,RNF
count,10000.00,10000.0,10000.00,10000.00,10000.00,10000.00,10000.00,10000.00,10000.00,10000.00,10000.00,10000.00
mean,5000.50,300.0,310.01,1538.78,39.99,107.95,0.03,0.00,0.01,0.01,0.01,0.00
std,2886.90,2.0,1.48,179.28,9.97,63.65,0.18,0.07,0.11,0.10,0.10,0.04
min,1.00,295.3,305.70,1168.00,3.80,0.00,0.00,0.00,0.00,0.00,0.00,0.00
25%,2500.75,298.3,308.80,1423.00,33.20,53.00,0.00,0.00,0.00,0.00,0.00,0.00
50%,5000.50,300.1,310.10,1503.00,40.10,108.00,0.00,0.00,0.00,0.00,0.00,0.00
75%,7500.25,301.5,311.10,1612.00,46.80,162.00,0.00,0.00,0.00,0.00,0.00,0.00
max,10000.00,304.5,313.80,2886.00,76.60,253.00,1.00,1.00,1.00,1.00,1.00,1.00


In [36]:
print('Categorical Feature Summery: ')
df.describe(include=['object'])

Categorical Feature Summery: 


,Product ID,Type
count,10000,10000
unique,10000,3
top,M14860,L
freq,1,6000


In [37]:
df['Type'].value_counts(normalize=True)


Type
L    0.6000
M    0.2997
H    0.1003
Name: proportion, dtype: float64

## 6.Categorical Column Exploration

In [49]:
# Machine Type Distribution
type_counts = df['Type'].value_counts().sort_index()
type_pct = df['Type'].value_counts(normalize=True).sort_index()*100

type_summery = pd.DataFrame({
    'Machine Type': type_counts.index,
    'Count': type_counts.values,
    'Percentage (%)': type_pct.values.round(2)
})

print(type_summery.to_string(index=False))
print(f'\nTotal unique machine types: {df["Type"].nunique()}')

Machine Type  Count  Percentage (%)
           H   1003           10.03
           L   6000           60.00
           M   2997           29.97

Total unique machine types: 3


In [56]:
# Product Id Exploration
print(f' Total Unique Product IDs: {df['Product ID'].nunique():,}')
print(f' Total Rows: {len(df):,}')
print(f'{df[['Product ID','Type']].head(10)}')    #Each Product ID is unique — it acts as an identifier, not a feature.

 Total Unique Product IDs: 10,000
 Total Rows: 10,000
  Product ID Type
0     M14860    M
1     L47181    L
2     L47182    L
3     L47183    L
4     L47184    L
5     M14865    M
6     L47186    L
7     L47187    L
8     M14868    M
9     M14869    M


## 7.Target Variable Analysis - Machine Failure

In [66]:
# Machine Falure Distribution
failure_counts = df['Machine failure'].value_counts().sort_index()
failure_pct = df['Machine failure'].value_counts(normalize=True).sort_index() * 100

print("Machine Failure Distribution: ")
failure_summery = pd.DataFrame({
    'Status': ['No Failure (0)','Failure (1)'],
    'Count': failure_counts.values,
    'Percentage (%)': failure_pct.values.round(2)
})

print(failure_summery.to_string(index=False))

Machine Failure Distribution: 
        Status  Count  Percentage (%)
No Failure (0)   9661           96.61
   Failure (1)    339            3.39


**Imbalance Data Detected** - Failures are only 3.39% of total data.

## 8.Failure Type Breakdown

In [72]:
# Failure Type Defination
failure_type_names = {
    'TWF': 'Tool Wear Failure',
    'HDF': 'Heat Dissipation Failure',
    'PWF': 'Power Failure',
    'OSF': 'Overstrain Failure',
    'RNF': 'Random Failure'
}

failure_types = ['TWF', 'HDF', 'PWF', 'OSF', 'RNF']
total_failures = df['Machine failure'].sum()

print("Failure Type Breakdown\n")

failure_df = pd.DataFrame({
    'Code': failure_types,
    'Failure Type': [failure_type_names[f] for f in failure_types],
    'Count': [df[f].sum() for f in failure_types],
    'Percentage of Total Failures (%)': [(df[f].sum() / total_failures * 100).round(2) for f in failure_types]
})

failure_df = failure_df.sort_values('Count', ascending=False).reset_index(drop=True)
print(failure_df.to_string(index=False))

Failure Type Breakdown

Code             Failure Type  Count  Percentage of Total Failures (%)
 HDF Heat Dissipation Failure    115                             33.92
 OSF       Overstrain Failure     98                             28.91
 PWF            Power Failure     95                             28.02
 TWF        Tool Wear Failure     46                             13.57
 RNF           Random Failure     19                              5.60


In [75]:
print(f'Total Failures: {int(total_failures)}')
print(f'Total Faiure Type Entries: {df[failure_types].sum().sum()}')

Total Failures: 339
Total Faiure Type Entries: 373


In [77]:
# check failure without specific types
no_type_failures = df[(df['Machine failure'] == 1) & (df['Machine failure'] == 0) ]

print("Failure without specific type")
print(f'Count {len(no_type_failures)}')

if len(no_type_failures) > 0:
    print(f'\n These are failures where no specific type is recorded.')
else:
    print('\n All failures have a specific failure type assigned.')


Failure without specific type
Count 0

 All failures have a specific failure type assigned.


---

## INITIAL OBSERVATIONS

---

### 1. Dataset Overview
- **Total records**    : 10,000  
- **Total features**   : 14  
- **Missing values**   : None   
- **Duplicate rows**   : 0  

---

### 2. Machine Types
- **3 machine types**:  
  - **L** — Low  
  - **M** — Medium  
  - **H** — High  
- Distribution is **roughly balanced** across machine types

---

### 3. Target Variable — Machine Failure
- **Total failures**   : 339  
- **Failure rate**     : 3.39%  
- **Severe class imbalance observed**  
  - Majority class (No Failure): ~96.61%  
  - Minority class (Failure): ~3.39%  
- This imbalance must be considered during modeling and evaluation

---

### 4. Failure Types
- **Five failure sub-types** identified:
  - TWF — Tool Wear Failure  
  - HDF — Heat Dissipation Failure  
  - PWF — Power Failure  
  - OSF — Overstrain Failure  
  - RNF — Random Failure  
- Some failure modes occur more frequently than others  
- **Root cause analysis** will be explored in later notebooks

---

### 5. Sensor & Operational Features
- **Environmental & process variables**:
  - Air temperature  
  - Process temperature  
- **Mechanical variables**:
  - Rotational speed  
  - Torque  
  - Tool wear  
- Correlation with machine failure will be analyzed in subsequent steps

---

## Next Steps — Notebook 02: Data Cleaning

- Rename columns for better clarity and consistency  
- Explicitly document class imbalance  
- Perform feature engineering (if required)  
- Load cleaned data into SQLite database  
- Save processed dataset for downstream analysis  

---
